client/src/components/ViewPostDialog.tsx

```diff
@@ -159,49 +159,54 @@ function LikeButton(props: { numLikes: number }) {
   );
 }
 
-function CapacityBar(props: { maxCapacity: number }) {
-  // TODO: This data should be synced with db -- models needs to be updated
-  const [capacity, setCapacity] = React.useState(0);
-  const [isCheckedin, toggleCheckin] = React.useState(false);
+function CapacityBar(props: {
+  maxCapacity: number;
+  postID: string;
+  isUserCheckedIn: boolean;
+  usersCheckedIn: number;
+}) {
   const maxCapacity = !isNaN(props.maxCapacity) ? props.maxCapacity : 0;
 
-  const handleCheckIn = () => {
-    toggleCheckin((prev) => !prev);
-  };
-
-  React.useEffect(() => {
-    if (isCheckedin) {
-      setCapacity((prev) => prev + 1);
+  const handleCheckIn = async () => {
+    if (props.isUserCheckedIn) {
+      await api.checkout(props.postID);
+      // As there is no global state management system, we would have to wait
+      // for the autoreload system to update the post info. This is a hack to
+      // ensure that the checked in state is immediately updated.
+      props.isUserCheckedIn = false;
+      props.usersCheckedIn -= 1;
     } else {
-      setCapacity((prev) => (prev > 0 ? prev - 1 : prev));
+      const result = await api.checkin(props.postID);
+      if (result.status !== 409) {
+        props.isUserCheckedIn = true;
+      }
+      // TODO indicate a standard alert to the user that the event could not be
+      // checked into (over capacity)
     }
-  }, [isCheckedin]);
+  };
 
-  const buttonHandler =
-    capacity < props.maxCapacity ? (
-      isCheckedin ? (
-        <Button onClick={handleCheckIn} variant='contained'>
-          Undo
-        </Button>
-      ) : (
-        <Button onClick={handleCheckIn} variant='outlined'>
-          Check In
-        </Button>
-      )
-    ) : (
-      <Button disabled variant='outlined'>
-        AT CAPACITY
-      </Button>
-    );
+  const buttonHandler = props.isUserCheckedIn ? (
```

Comment:
Add some tests to make sure this behaves correctly. You can check `PostDashboard.test.tsx` and do a similar method of mocking post responses.

Reply:
I think for something like this its fine as is.